plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java

```diff
@@ -716,4 +773,80 @@ public String getName()
             return name;
         }
     }
+
+    private static class RemoteDatabaseEventMonitor
+            implements Runnable
+    {
+        private static final Logger log = Logger.get(RemoteDatabaseEventMonitor.class);
+
+        private final Jdbi jdbi;
+        private final Set<RemoteLogTracingEvent> tracingEvents;
+        private ScheduledThreadPoolExecutor executor;
+
+        private RemoteDatabaseEventMonitor()
+        {
+            jdbi = Jdbi.create(JDBC_URL, JDBC_USER, JDBC_PASSWORD);
+            tracingEvents = Sets.newConcurrentHashSet();
+        }
+
+        public void startTracingDatabaseEvent(RemoteLogTracingEvent event)
+        {
+            if (tracingEvents.isEmpty()) {
+                executor = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("redshift-database-event-monitor"));
+                executor.scheduleWithFixedDelay(this, 0, 5, SECONDS);
+            }
+            tracingEvents.add(event);
+        }
+
+        public void stopTracingDatabaseEvent(RemoteLogTracingEvent event)
+        {
+            tracingEvents.remove(event);
+            if (tracingEvents.isEmpty()) {
+                executor.shutdown();
+            }
+        }
+
+        @Override
+        public void run()
+        {
+            if (tracingEvents.isEmpty()) {
+                return;
+            }
+
+            try {
+                getRecentQueries()
+                        .forEach(remoteDatabaseEvent -> tracingEvents.forEach(tracingEvent -> tracingEvent.accept(remoteDatabaseEvent)));
+            }
+            catch (Exception e) {
+                // ignore exceptions to keep scheduled executions going
+                log.warn(e, "Encountered error while gathering Redshift remote database events");
+            }
+        }
+
+        private List<RemoteDatabaseEvent> getRecentQueries()
+        {
+            try (Handle handle = jdbi.open()) {
+                return handle.createQuery("""
+                                SELECT query_text, status, error_message
+                                FROM SYS_QUERY_HISTORY
+                                WHERE database_name = :db_name
+                                AND query_type = 'SELECT'
+                                AND user_id = current_user_id
+                                AND start_time > GETDATE() - INTERVAL '15 minutes'
+                                """)
+                        .bind("db_name", TEST_DATABASE)
+                        .map((rs, _) -> new RemoteDatabaseEvent(
+                                rs.getString("query_text"),
+                                switch (requireNonNull(rs.getString("status"), "status is null").trim()) {
+                                    case "failed" -> Optional.ofNullable(rs.getString("error_message"))
+                                            .flatMap(message -> message.contains("cancelled on user's request") ? Optional.of(Status.CANCELLED) : Optional.empty())
```

Comment:
Can we assume a status here instead of Optional.empty() when there is no cancellation for case `failed` ?